hazelcast/include/hazelcast/client/imap.h

```diff
@@ -120,8 +128,36 @@ namespace hazelcast {
             * @return the previous value. if there is no mapping for key, then returns boost::none.
             */
             template<typename K, typename V, typename R=V>
-            boost::future<boost::optional<R>> put(const K &key, const V &value, std::chrono::milliseconds ttl) {
-                return to_object<R>(put_internal(to_data(key), to_data(value), ttl));
+            boost::future<boost::optional<R>>
+            put(const K &key, const V &value, std::chrono::milliseconds ttl) {
+                using boost::launch;
+                using boost::future;
+                using serialization::pimpl::data;
+                struct state_ {
+                  future<data> key_data_fut;
```

Comment:
Did you consider using `boost::when_all` or `boost::when_any`. E.g.:
```
            template<typename K, typename V, typename R=V>
            boost::future<boost::optional<R>>
            put(const K &key, const V &value, std::chrono::milliseconds ttl) {
                using boost::launch;
                using boost::future;
                using serialization::pimpl::data;
                return boost::when_all(to_data2(key), to_data2(value)).then(sync, [this, ttl] (std::vector<future<data>> futures) {
                    return std::make_pair(futures[0].get(), futures[1].get());
                }).then(sync, [this, ttl] (future<std::pair<data, data>> f){
                    auto data_pair  = f.get();
                    return put_internal(data_pair.first, data_pair.second, ttl);
                }).unwrap();
            }
```

I still don't want to use method `unwrap` which is experimental, there may be a way to even eliminate that.

Reply:
when_all opens a thread that is why I avoided it.

Reply:
> I still don't want to use method unwrap which is experimental, there may be a way to even eliminate that.

The `boost::future::then` is also experimental: https://www.boost.org/doc/libs/1_78_0/doc/html/thread/synchronization.html#thread.synchronization.futures.reference.unique_future.then